src/mlpack/core/data/dictionary_encoding.hpp

```diff
@@ -0,0 +1,135 @@
+/**
+ * @file dictionary_encoding.hpp
+ * @author Jeffin Sam
+ *
+ * Implementation of dictionary encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_DICT_ENCODING_HPP
+#define MLPACK_CORE_DATA_DICT_ENCODING_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace data {
+/**
+ * A simple Dictionary Enocding class
+ */
+class DicitonaryEncoding
+{
+ public:
+  /**
+  * A function to create map using a corpus.
+  * 
+  * @param strings Corpus of text to encode.
+  * @param delimiter Delimiator use to split the corpus.
+  */
+  void CreateMap(std::string& strings, std::string delimiter = " ");
+
+  /**
+  * Overloaded function to create map using a corpus, which
+  * allowes users to provide their own custom tokenization.
+  * 
+  * @param strings Corpus of text to encode.
+  * @param delimiter Delimiator use to split the corpus.
+  * @param tokenizer A lamda function providing the rule for tokenization.
+  */
+  template<typename TokenizerType>
+  void CreateMap(std::string& strings, std::string delimiter,
+                 TokenizerType tokenizer);
+
+  /**
+  * A function to reset the mapping that is clear all the encodings
+  */
+  void Reset();
+
+  /**
+  * A fucntion to encode given array of strings using a particular delimiter,
+  * providing custom rule for tokenization.
+  *
+  * For example 
+  * Vector is :
+  * [hello@wow, wow@hello@good] would be encoded using '@' as delimiter as 
+  * [1 2 0, 2 1 3] 
+  * The function paddes 0 to maintain same sizes across all the rows
+  * User may also provide their custom tokenization rule.
+  *
+  * @param strings Vector of strings.
+  * @param delimiter Delimiter used to split the strings.
+  * @param tokenizer A lamda function providing the rule for tokenization.
+  * @param output Output Matrix to store encoded results.
+  */
+  template<typename MatType, typename TokenizerType>
+  void DictEncode(const std::vector<std::string>& strings,
+                  const std::string& delimiter,
+                  TokenizerType tokenizer,
```

Comment:
I guess you needn't a delimiter if you use `TokenizerType`. My original point is that `TokenizerType` defines the tokenization rules entirely. For example it can contain the delimiter inside itself.

Reply:
Left some message for you on IRC. Sorry for not clearing my plan of API up to you. I was looking for more of user friendly interface. My plan was to handle small and less complex delimiters internally such as `" "` or `"-"` or `"@"` using the function `dictencode(string,delimiter)` and if user was looking for more complex tokenization then he can always use the function object and pass it as an argument to `dictencode(string,delimiter,tokenizer)`

But I guess , your point of providing full customization may be apt since it gives full control to user.

Have made apt changes.

Reply:
I think it is better to write some standard tokenizers to handle these special cases. So, the user could either choose a standard tokenizer or write its own.

Reply:
Yes I agree :) . Added a standard tokenizer .